### PR TITLE
[refactor] Rearrange taichi/common and taichi/util. taichi/core is removed.

### PR DESCRIPTION
**Pull request by yuanming-hu**
Not sure it makes sense or not, but when I try to include a file in one of the three folders, I often forget which folder to use. The three folds are essentially doing the same thing...

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by archibate** [comment]
Great but I personally think util is not that usage, it's about `image_io.h` etc., maybe `bit.h` is ok, but `commit_hash.h` is not ok, it's not a util at all.

**Comment by k-ye** [comment]
Yeah, I also feel that `common` means some base components that are widely used by the repo, whereas `utils` are some non-core helpers..? At a minimum, since `core` only contains `.cpp`, is it possible to put these `cpp`s to each of their corresponding header's directory?

**Comment by yuanming-hu** [comment]
Thanks for the inputs! How about this:
 - `core` has only cpp files, so let's split this folder into `common` and `util` (as suggested by @k-ye)
 - Merge `core/interface` (empty cpp) and `core/interfaces`
 - Rename `common/util.h` -> `common/core.h`
 - Move `common/bit.h` (thanks to @archibate),  `common/testing` (not sure about this one, but this is only used in C++ tests) to `util` 
   - Remaining files in `common` would be `dict` (which will be removed later since it's no longer used frequently), `interface.h`, `serialization.h`, `common/meta`, `task.h`, `util.h`(will be renamed to `core.h`)

**Comment by archibate** [comment]
Ack :)

> Rename common/util.h -> common/core.h

but could you remove functions like `split_string`, `starts_with`, etc. to `util/xxx.h`? We want only the core compoment left in `core.h`.

**Comment by yuanming-hu** [comment]
> Ack :)
> 
> > Rename common/util.h -> common/core.h
> 
> but could you remove functions like `split_string`, `starts_with`, etc. to `util/string_utils.h`? We want only the core compoment left in `core.h`.

Sounds good. I would do that in a later PR so that this one has only file movements.

File: taichi/common/dict.h [hunk 1]
```diff
@@ -14,8 +14,8 @@
 #include <sstream>
 #include <typeinfo>
 
+#include "taichi/common/core.h"
 #include "taichi/math/math.h"
-#include "util.h"
```
Comment (archibate) [inline taichi/common/dict.h:18]: dict? Move this file to util?

File: taichi/common/core.h [hunk 2]
```diff
@@ -264,7 +264,7 @@ TI_NAMESPACE_END
 //                           Meta-programming
 //******************************************************************************
 
-#include "meta.h"
+#include "taichi/util/meta.h"
```
Comment (archibate) [inline taichi/common/core.h:267]: Good, don't forget to remove these stuff.

File: taichi/common/dict.h (hunk 1 above)
Comment (yuanming-hu) [reply taichi/common/dict.h:18]: As mentioned above `dict` will be removed later.

File: taichi/common/core.h (hunk 2 above)
Comment (yuanming-hu) [reply taichi/common/core.h:267]: Thanks. As mentioned above these will be done in a later PR so that this one has only file movements.

**Pull request merged**